replication-api/src/main/java/org/codice/ditto/replication/api/modern/Persistable.java

```diff
@@ -0,0 +1,71 @@
+/**
+ * Copyright (c) Connexta
+ *
+ * <p>This is free software: you can redistribute it and/or modify it under the terms of the GNU
+ * Lesser General Public License as published by the Free Software Foundation, either version 3 of
+ * the License, or any later version.
+ *
+ * <p>This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details. A copy of the GNU Lesser General Public
+ * License is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+package org.codice.ditto.replication.api.modern;
+
+import java.util.Map;
+import org.apache.commons.lang3.reflect.ConstructorUtils;
+
+/**
+ * A Persistable is an object that can be saved in the {@link ReplicationPersistentStore}. Any
+ * Persistable must have a persistence type, and a map of properties including an ID that will be
+ * saved in storage.
+ */
+public interface Persistable {
+
+  /**
+   * Returns the ID of the persistable object
+   *
+   * @return The ID of the object
+   */
+  String getId();
+
+  /**
+   * Returns the persistence type of the persistable object
+   *
+   * @return The persistence type of the object
+   */
+  String getType();
+
+  /**
+   * Returns the map of properties that define the persistable object. This will also contain the ID
+   *
+   * @return The properties of the object
+   */
+  Map<String, Object> getProperties();
+
+  /**
+   * A method that allows the client to flexibly create a persistable from the given map. In order
+   * for this method to work every class that implements the Persistable interface needs to have a
+   * constructor that takes in a {@code Map<String, Object>} as its sole argument. Otherwise a call
+   * to this method with that class as an argument will throw an {@link IllegalArgumentException}.
+   *
+   * @param klass The class of object to create an instance of.
+   * @param properties The properties that the created object will have.
+   * @param <T> A class implementing the persistable interface, whose constructor will be called and
+   *     the created object returned.
+   * @return An instance of T that has been created with the given properties.
+   */
+  static <T extends Persistable> T createPersistable(

```

Comment:
This seems weird because the caller is required to know an implementation of Persistable. I can't just fetch a `ReplicationSite` right? I have to fetch a `ReplicationSiteImpl`. It seems like it should be `ReplicationSite implements Persistable`

Reply:
❗️ Agreed. Such a method should not be in the API at all. It should be defined in a service and allow the implementation to register the service. In addition, identification should be done via the object's interface and not the implementation's class.